package/libpiksi/libpiksi/src/settings.c

```diff
@@ -471,30 +752,163 @@ static int setting_register(settings_ctx_t *ctx, const char *section,
+ * @param ctx: setting context
+ * @param setting_data: setting to read from settings daemon
+ * @return zero on success, -1 the transaction failed to complete
+ */
+static int setting_read_watched_value(settings_ctx_t *ctx, setting_data_t *setting_data)
+{
+  /* Build message */
+  u8 msg[SBP_PAYLOAD_SIZE_MAX];
+  u8 msg_len = 0;
+  int l;
 
-  compare_deinit(ctx);
+  if (!setting_data->watchonly) {
+    piksi_log(LOG_ERR, "cannot update non-watchonly setting manually");
+    return -1;
+  }
 
-  if (!success) {
-    piksi_log(LOG_ERR, "setting registration failed");
+  l = message_header_get(setting_data, &msg[msg_len], sizeof(msg) - msg_len);
+  if (l < 0) {
+    piksi_log(LOG_ERR, "error building settings read req message");
     return -1;
   }
+  msg_len += l;
+
+  return setting_perform_request_reply_from(ctx,
+                                            SBP_MSG_SETTINGS_READ_REQ,
+                                            msg,
+                                            msg_len,
+                                            msg_len,
+                                            WATCH_INIT_TIMEOUT_MS,
+                                            WATCH_INIT_TRIES,
+                                            SBP_SENDER_ID);
+}
+
+/**
+ * @brief setting_parse_setting_text - parse main components of settings payload
+ * @param msg: raw sbp message
+ * @param msg_n: length of sbp message
+ * @param section: reference to location of section string in message
+ * @param name: reference to location of name string in message
+ * @param value: reference to location of value string in message
+ * @return zero on successful parse, -1 on parsing error
+ */
+static int setting_parse_setting_text(const u8 *msg,
+                                      u8 msg_n,
+                                      const char **section,
+                                      const char **name,
+                                      const char **value)
+{
+  const char **result_holders[] = { section, name, value };
+  u8 start = 0;
+  u8 end = 0;
+  for (u8 i = 0; i < sizeof(result_holders) / sizeof(*result_holders); i++) {
+    bool found = false;
+    *(result_holders[i]) = NULL;
+    while (end < msg_n) {
+      if (msg[end] == '\0') {
+        if (end == start) {
+          return -1;
+        } else {
+          *(result_holders[i]) = (const char *)msg + start;
+          start = (u8)(end + 1);
+          found = true;
+        }
+      }
+      end++;
+      if (found) {
+        break;
+      }
+    }
+  }
+  return 0;
+}
+
+/**
+ * @brief setting_format_setting - formats a fully formed setting message payload
+ * @param setting_data: the setting to format
+ * @param buf: buffer to hold formatted setting string
+ * @param len: length of the destination buffer
+ * @return bytes written to the buffer, -1 in case of failure
+ */
+static int setting_format_setting(setting_data_t *setting_data, char *buf, int len)
```

Comment:
Can we make this a library function?

Reply:
I'll try but this one is tricky since the setting_data struct is opaque. I would need to expose some way for the user to obtain a reference to one in order for it to be useful outside of the module internally. Not entirely certain on the benefit but I'm open to suggestions if you had a vision for it.

Reply:
Never mind then, parse is more important than formatting, I guess what I'm ultimately looking for is a high level API to parse/format settings data (to use in things like the settings write whitelist), and then whatever we need to build a command line tool for reading/writing settings.  But if these things are not easy to add right now, we can tackle them later.

Reply:
Ok cool, yeah I had originally envisioned these getting pulled into libsbp as we iterate on refactoring the settings code. This was really just meant to be a first pass, there is still a bunch to do!